homeassistant/components/octoprint/config_flow.py

```diff
@@ -0,0 +1,198 @@
+            return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self._create_setup_failure(user_input, error)
+
+    async def async_step_auth_failed(self, user_input):
+        """Handle api fetch failure."""
+        return self._create_setup_failure(self.user_input, "auth_failed")
+
+    async def async_step_import(self, user_input):
+        """Handle import."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle discovery flow."""
+        uuid = discovery_info["properties"]["uuid"]
+        await self.async_set_unique_id(uuid)
+        self._abort_if_unique_id_configured()
+
+        self.context["title_placeholders"] = {
+            CONF_HOST: discovery_info[CONF_HOST],
+        }
+
+        self.discovery_schema = _schema_with_defaults(
+            host=discovery_info[CONF_HOST],
+            port=discovery_info[CONF_PORT],
+            path=discovery_info["properties"][CONF_PATH],
+        )
+
+        return await self.async_step_user()
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle ssdp discovery flow."""
+        uuid = discovery_info["UDN"][5:]
+        await self.async_set_unique_id(uuid)
+        self._abort_if_unique_id_configured()
+
+        url = urlsplit(discovery_info["presentationURL"])
+        self.context["title_placeholders"] = {
+            CONF_HOST: url.hostname,
+        }
+
+        self.discovery_schema = _schema_with_defaults(
+            host=url.hostname,
+            port=url.port,
+        )
+
+        return await self.async_step_user()
+
+    async def _async_get_auth_key(self):
+        """Get application api key."""
+        session = async_get_clientsession(self.hass)
+        octoprint = OctoprintClient(
+            self.user_input[CONF_HOST],
+            session,
+            self.user_input[CONF_PORT],
+            self.user_input[CONF_SSL],
+            self.user_input[CONF_PATH],
+        )
+
+        try:
+            self.user_input[CONF_API_KEY] = await octoprint.request_app_key(
+                "Home Assistant", self.user_input[CONF_USERNAME], 30
+            )
+        finally:
+            # Continue the flow after show progress when the task is done.
+            self.hass.async_create_task(
+                self.hass.config_entries.flow.async_configure(
+                    flow_id=self.flow_id, user_input=self.user_input
+                )
+            )
+
+    def _create_setup_failure(self, user_input: dict, error_code: str):
+        data_schema = _schema_with_defaults(
+            username=user_input[CONF_USERNAME],
+            host=user_input[CONF_HOST],
+            port=user_input[CONF_PORT],
+            path=user_input[CONF_PATH],
+            ssl=user_input[CONF_SSL],
+        )
+        return self.async_show_form(
+            step_id="user", data_schema=data_schema, errors={"base": error_code}
```

Comment:
Don't move directly to another step's form.

Reply:
What would the correct way to move back to the user form be at this point?